### PR TITLE
Rewrite IREquality to use a more compact stack instead of deep recursion

### PR DESCRIPTION
**Pull request by abadams**
Deletes a bunch of code and speeds up lowering time of local laplacian with 20 pyramid levels by ~2.5%

Also fix some cases where equal wasn't considering recently-added IR node fields like partition_policy (suggestions for some way to guarantee this can't happen would be welcome).

Also changed calls to equal(..., true) to just is_const_one

### CODE REVIEW

**Comment by abadams** [comment]
Now more like 4% faster, and fast enough to delete the specialized version of equal in IRMatch.cpp

File: src/IREquality.cpp [hunk 1]
```diff
@@ -10,725 +10,558 @@ using std::vector;
...
+        }
     }
 
-    if (ha == nullptr) {
-        // void* < T*
-        result = LessThan;
-        return result;
-    }
+    size_t hash(const IRNode *a, const IRNode *b) {
+        uintptr_t pa = (uintptr_t)a;
```
Comment (steven-johnson) [inline src/IREquality.cpp:58]: Would be nice to comment what hash algo this is for future reference

File: src/IREquality.cpp [hunk 2]
```diff
@@ -10,725 +10,558 @@ using std::vector;
...
+        } else if (!ha) {
+            result = Order::LessThan;
+        } else if (!hb) {
+            result = Order::GreaterThan;
+        } else {
+            // They're both non-void handle types with distinct type info
+            // structs. We now need to distinguish between different C++
+            // pointer types (e.g. char * vs const float *). If would be nice
+            // if the structs were unique per C++ type. Then comparing the
+            // pointers above would be sufficient.  Unfortunately, different
```
Comment (steven-johnson) [inline src/IREquality.cpp:154]: Highly useful comment!

File: src/IREquality.cpp [hunk 3]
```diff
@@ -10,725 +10,558 @@ using std::vector;
...
     }
 
-    for (size_t i = 0; i < ha->cpp_type_modifiers.size(); i++) {
-        compare_scalar(ha->cpp_type_modifiers[i],
-                       hb->cpp_type_modifiers[i]);
+    HALIDE_ALWAYS_INLINE
+    void cmp(double a, double b) {
+        // Floating point scalars need special handling, due to NaNs.
+        if (std::isnan(a) && std::isnan(b)) {
+        } else if (std::isnan(a)) {
```
Comment (steven-johnson) [inline src/IREquality.cpp:194]: Did you mean to leave the both-nan case empty? If so, please add a comment like `/*deliberately empty, leave comparison state unchanged */`

**Review by steven-johnson — APPROVED** [review]
LGTM pending comments

**Comment by abadams** [comment]
Merging. Bots were all green a few commits ago and I've only tweaked comments since then.

**Pull request merged**